src/mapping/nav2_map_server/src/map_reps/occupancy_grid.cpp

```diff
@@ -0,0 +1,313 @@
+#include <cstring>
+#include <fstream>
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+#include "yaml-cpp/yaml.h"
+
+#define MAP_IDX(sx, i, j) ((sx) * (j) + (i))
+
+
+template<typename T>
+void operator>>(const YAML::Node & node, T & i)
+{
+  i = node.as<T>();
+}
+
+// Interface //
+
+void OccGridLoader::loadMapInfoFromFile(std::string fname)
+{
+  std::ifstream fin(fname.c_str());
+  if (fin.fail()) {
+    fprintf(stderr,
+      "[ERROR] [map_server]: Map_server could not open %s\n", fname.c_str());
+    exit(-1);
+  }
+
+  YAML::Node doc = YAML::LoadFile(fname);
+
+  try {
+    doc["resolution"] >> res;
+  } catch (YAML::InvalidScalar) {
+    fprintf(stderr,
+      "[ERROR] [map_server]: The map does not contain a resolution tag or it is invalid.\n");
+    exit(-1);
+  }
+  try {
+    doc["negate"] >> negate;
+  } catch (YAML::InvalidScalar) {
+    fprintf(stderr,
+      "[ERROR] [map_server]: The map does not contain a negate tag or it is invalid.\n");
+    exit(-1);
+  }
+  try {
+    doc["occupied_thresh"] >> occ_th;
+  } catch (YAML::InvalidScalar) {
+    fprintf(stderr,
+      "[ERROR] [map_server]: The map does not contain an occupied_thresh tag or it is invalid.\n");
+    exit(-1);
+  }
+  try {
+    doc["free_thresh"] >> free_th;
+  } catch (YAML::InvalidScalar) {
+    fprintf(stderr,
+      "[ERROR] [map_server]: The map does not contain a free_thresh tag or it is invalid.\n");
+    exit(-1);
+  }
+  try {
+    std::string modeS = "";
+    doc["mode"] >> modeS;
+    if (modeS == "trinary") {
+      mode = TRINARY;
+    } else if (modeS == "scale") {
+      mode = SCALE;
+    } else if (modeS == "raw") {
+      mode = RAW;
+    } else {
+      fprintf(stderr,
+        "Invalid mode tag \"%s\".\n", modeS.c_str());
+      exit(-1);
+    }
+  } catch (YAML::Exception) {
+    fprintf(stdout,
+      "[DEBUG] [map_server] The map does not contain a mode tag or it is invalid.\n");
+    mode = TRINARY;
+  }
+  try {
+    doc["origin"][0] >> origin[0];
+    doc["origin"][1] >> origin[1];
+    doc["origin"][2] >> origin[2];
+  } catch (YAML::InvalidScalar) {
+    fprintf(stderr,
+      "[ERROR] [map_server]: The map does not contain an origin tag or it is invalid.\n");
+    exit(-1);
+  }
+  try {
+    doc["image"] >> mapfname;
+    // TODO(bpwilcox): make this path-handling more robust
+    if (mapfname.size() == 0) {
+      fprintf(stderr,
+        "[ERROR] [map_server]: The image tag cannot be an empty string.\n");
+      exit(-1);
+    }
+    if (mapfname[0] != '/') {
+      // dirname can modify what you pass it
+      char * fname_copy = strdup(fname.c_str());
+      mapfname = std::string(dirname(fname_copy)) + '/' + mapfname;
+      free(fname_copy);
+    }
+  } catch (YAML::InvalidScalar) {
+    fprintf(stderr,
+      "[ERROR] [map_server]: The map does not contain an image tag or it is invalid.\n");
+    exit(-1);
+  }
+}
+
+void OccGridLoader::loadMapFromFile(std::string mapfname)
+{
+  SDL_Surface * img;
```

Comment:
This can be done later during refactoring, but these variables should be moved to where they're first used.

Reply:
I'd go a bit further and say that the function is too long. You could push a lot of the sections of code to sub-functions.